### PR TITLE
feat: add nsfw field to application commands

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Setting `nsfw=True` disables/hides a command outside of nsfw channels - like many other fields, it cannot be set per subcommand or group.
Not announced/documented yet and support in DMs is still tbd/wip, hence the draft PR.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/app_commands.py [hunk 1]
```diff
@@ -437,13 +437,20 @@ class ApplicationCommand(ABC):
         Defaults to ``True``.
 
         .. versionadded:: 2.5
+
+    nsfw: :class:`bool`
+        Whether this command can only be used in NSFW channels.
+        Defaults to ``False``.
```
Comment (onerandomusername) [inline disnake/app_commands.py:443]: we might be able to merge this without dm support but we do not yet know how DMs will work.

Comment (lol219) [reply disnake/app_commands.py:443]: They don't show those commands in dms

**Review by lol219 — APPROVED** [review]
I guess it's good rn

**Comment by onerandomusername** [comment]
@shiftinv would you please add a changelog entry?

**Comment by shiftinv** [comment]
> @shiftinv would you please add a changelog entry?

https://github.com/DisnakeDev/disnake/pull/528/commits/54cbd12cdf8da764d4032ad31a13b3665821cd16

File: disnake/app_commands.py [hunk 2]
```diff
@@ -661,6 +685,7 @@ def from_dict(cls, data: ApplicationCommandPayload) -> Self:
             name=Localized(data["name"], data=data.get("name_localizations")),
             dm_permission=data.get("dm_permission") is not False,
             default_member_permissions=_get_as_snowflake(data, "default_member_permissions"),
+            nsfw=data.get("nsfw"),
```
Comment (onerandomusername) [inline disnake/app_commands.py:688]: should this default to False here, instead of None?

Comment (shiftinv) [reply disnake/app_commands.py:688]: It'll default to False in the constructor:
https://github.com/DisnakeDev/disnake/blob/54cbd12cdf8da764d4032ad31a13b3665821cd16/disnake/app_commands.py#L500

**Review by onerandomusername — APPROVED** [review]
looks good.

**Comment by onerandomusername** [comment]
@shiftinv thoughts on merging this even without discord confirming the feature?

**Comment by shiftinv** [comment]
> @shiftinv thoughts on merging this even without discord confirming the feature?

Should be fine, since there's a server-side check as well and it doesn't rely on client support like some other features.
It's not known yet if the field is nullable - I typed it as the most permissive type (optional+nullable) for now, so we won't run into issues.

**Pull request merged**